allennlp/nn/util.py

```diff
@@ -667,12 +685,51 @@ def sequence_cross_entropy_with_logits(logits: torch.FloatTensor,
         raise ValueError("Got average f{average}, expected one of "
                          "None, 'token', or 'batch'")
 
+    # make sure weights are float
+    weights = weights.float()
     # shape : (batch * sequence_length, num_classes)
     logits_flat = logits.view(-1, logits.size(-1))
     # shape : (batch * sequence_length, num_classes)
     log_probs_flat = torch.nn.functional.log_softmax(logits_flat, dim=-1)
     # shape : (batch * max_len, 1)
     targets_flat = targets.view(-1, 1).long()
+    # focal loss coefficient
+    if gamma:
+        # shape : (batch * sequence_length, num_classes)
+        probs_flat = log_probs_flat.exp()
+        # shape : (batch * sequence_length,)
+        probs_flat = torch.gather(probs_flat, dim=1, index=targets_flat)
+        # shape : (batch * sequence_length,)
+        focal_factor = (1. - probs_flat) ** gamma
+        focal_factor = focal_factor.view(-1, 1)
+    else:
```

Comment:
I think you can similarly modify `weights` as the last line of the `if` above, and not need this `else`, or need to reference `focal_factor` down below.

Reply:
Great. Update.

Reply:
Hi @matt-gardner. I realize there is a problem here because when calculating the average, "token", "batch", or even "none", it relies on `weights.sum()`. Do you think `gamma` (and also `alpha`) should go into the denominator of average `weights.sum()`? It is a kind of design decision, rather than correctness, of how to calculate the average here. What do you think?

Reply:
`alpha` and `gamma` have the effect of changing the weighting on each of the instances / tags in the batch, so it will affect the direction of the gradients.  Changing the denominator in `weights.sum()` changes all directions of the gradient equally, so including `alpha` and `gamma` in the sum effectively just lowers the learning rate.  Right?  If I'm understanding this correctly, I'd probably compute the sum before multiplying by the `alpha` and `gamma` factors, so the effective learning rate doesn't change as much.

Reply:
Right. The outcome would be changing the learning rage if they are included. If this is the case, I can calculate the `sum()` before involving them.